subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/repositories/resolver/ComponentDependenciesMetadataDetailsAdapter.java

```diff
@@ -0,0 +1,122 @@
+import java.util.List;
+import java.util.Map;
+
+public class ComponentDependenciesMetadataDetailsAdapter extends AbstractList<ComponentDependencyMetadata> implements ComponentDependenciesMetadataDetails {
+    private final List<DependencyMetadata> dependenciesMetadata;
+    private final Map<Integer, ComponentDependencyMetadata> componentDependencyMetadataAdapters;
+    private final NotationParser<Object, ComponentDependencyMetadataDetails> dependencyMetadataNotationParser;
+
+    public ComponentDependenciesMetadataDetailsAdapter(List<DependencyMetadata> dependenciesMetadata, NotationParser<Object, ComponentDependencyMetadataDetails> dependencyMetadataNotationParser) {
+        this.dependenciesMetadata = dependenciesMetadata;
+        this.componentDependencyMetadataAdapters = Maps.newHashMap();
+        this.dependencyMetadataNotationParser = dependencyMetadataNotationParser;
+    }
+
+    @Override
+    public ComponentDependencyMetadata get(int index) {
+        if (!componentDependencyMetadataAdapters.containsKey(index)) {
+            componentDependencyMetadataAdapters.put(index, new ComponentDependencyMetadataAdapter(dependenciesMetadata.get(index)));
+        }
+        return componentDependencyMetadataAdapters.get(index);
+    }
+
+    @Override
+    public int size() {
+        return dependenciesMetadata.size();
+    }
+
+    @Override
+    public ComponentDependencyMetadata remove(int index) {
+        ComponentDependencyMetadata componentDependencyMetadata = get(index);
+        dependenciesMetadata.remove(index);
+        componentDependencyMetadataAdapters.remove(index);
+        return componentDependencyMetadata;
+    }
+
+    @Override
+    public void add(String dependencyNotation) {
+        doAdd(dependencyNotation, null);
+    }
+
+    @Override
+    public void add(Map<String, String> dependencyNotation) {
+        doAdd(dependencyNotation, null);
+    }
+
+    @Override
+    public void add(String dependencyNotation, Action<ComponentDependencyMetadataDetails> configureAction) {
+        doAdd(dependencyNotation, configureAction);
+    }
+
+    @Override
+    public void add(Map<String, String> dependencyNotation, Action<ComponentDependencyMetadataDetails> configureAction) {
+        doAdd(dependencyNotation, configureAction);
+    }
+
+    private void doAdd(Object dependencyNotation, @Nullable Action<ComponentDependencyMetadataDetails> configureAction) {
+        ComponentDependencyMetadataDetails componentDependencyMetadataDetails = dependencyMetadataNotationParser.parseNotation(dependencyNotation);
+        if (configureAction != null) {
+            configureAction.execute(componentDependencyMetadataDetails);
+        }
+        maybeRemoveExistingDependency(componentDependencyMetadataDetails);
+        dependenciesMetadata.add(toDependencyMetadata(componentDependencyMetadataDetails));
+    }
+
+    private void maybeRemoveExistingDependency(ComponentDependencyMetadataDetails newDependency) {
```

Comment:
⭕️ Is this necessary? What do we do when you add the "same" dependency multiple times in the `project.dependencies {}` block?

Reply:
The `project.dependencies {}` allows multiple declaration and the resolution will pick one. E.g. latest one or the one that is _forced_.

Since it usually makes no sense to declare a dependency twice with different version ranges, I was thinking the semantics that `add()` replaces the existing one is the most intuitive here.

The use case would be that I want to set a certain dependency no matter what was defined before. Otherwise, if there is another specification of that dependency already (e.g. with a higher version), the results of the resolution might be confusing.

But the more I think about it: Maybe it is better to keep this in sync with the behaviour in `project.dependencies {}`. What do you think?